client/internal/httpsender_test.go

```diff
@@ -98,3 +100,50 @@ EKTcWGekdmdDPsHloRNtsiCa697B2O9IFA==
 
 	return cert, nil
 }
+
+func TestHTTPSenderRetryForFailedRequests(t *testing.T) {
+
+	srv := StartMockServer(t)
+	srv.srv.Config.ReadTimeout = 1 * time.Nanosecond
+	var buf []byte
+	srv.OnRequest = func(w http.ResponseWriter, r *http.Request) {
+		buf, _ = io.ReadAll(r.Body)
+		w.WriteHeader(http.StatusOK)
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	url := "http://" + srv.Endpoint
+	sender := NewHTTPSender(&sharedinternal.NopLogger{})
+	sender.NextMessage().Update(func(msg *protobufs.AgentToServer) {
+		msg.AgentDescription = &protobufs.AgentDescription{
+			IdentifyingAttributes: []*protobufs.KeyValue{{
+				Key: "service.name",
+				Value: &protobufs.AnyValue{
+					Value: &protobufs.AnyValue_StringValue{StringValue: "test-service"},
+				},
+			}},
+		}
+	})
+	sender.callbacks = types.CallbacksStruct{
+		OnConnectFunc: func() {
+		},
+		OnConnectFailedFunc: func(_ error) {
+		},
+	}
+	sender.url = url
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		sender.sendRequestWithRetries(ctx)
+		wg.Done()
+	}()
+	go func() {
+		time.Sleep(200 * time.Millisecond)
+		srv.srv.Config.ReadTimeout = 10 * time.Second
```

Comment:
This seems like a possibility for data race. Something is reading this field during the request.

Reply:
line at 107 also.

```
==================
WARNING: DATA RACE
Write at 0x00c0002dc030 by goroutine 20:
  github.com/open-telemetry/opamp-go/client/internal.TestHTTPSenderRetryForFailedRequests()
      /Users/srikanthccv/Github/opamp-go/client/internal/httpsender_test.go:107 +0x78
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.20.2/libexec/src/testing/testing.go:1576 +0x180
  testing.(*T).Run.func1()
      /opt/homebrew/Cellar/go/1.20.2/libexec/src/testing/testing.go:1629 +0x40

Previous read at 0x00c0002dc030 by goroutine 21:
  net/http.http2ConfigureServer()
      /opt/homebrew/Cellar/go/1.20.2/libexec/src/net/http/h2_bundle.go:3999 +0x174
  net/http.(*Server).onceSetNextProtoDefaults()
      /opt/homebrew/Cellar/go/1.20.2/libexec/src/net/http/server.go:3331 +0x12c
  net/http.(*Server).onceSetNextProtoDefaults_Serve()
      /opt/homebrew/Cellar/go/1.20.2/libexec/src/net/http/server.go:3312 +0x100
  net/http.(*Server).onceSetNextProtoDefaults_Serve-fm()
      <autogenerated>:1 +0x34
  sync.(*Once).doSlow()
      /opt/homebrew/Cellar/go/1.20.2/libexec/src/sync/once.go:74 +0xb0
  sync.(*Once).Do()
      /opt/homebrew/Cellar/go/1.20.2/libexec/src/sync/once.go:65 +0x40
  net/http.(*Server).setupHTTP2_Serve()
      /opt/homebrew/Cellar/go/1.20.2/libexec/src/net/http/server.go:3306 +0x230
  net/http.(*Server).Serve()
      /opt/homebrew/Cellar/go/1.20.2/libexec/src/net/http/server.go:3038 +0x24c
  net/http/httptest.(*Server).goServe.func1()
      /opt/homebrew/Cellar/go/1.20.2/libexec/src/net/http/httptest/server.go:310 +0xa0
```